File: apps/desktop/src/lib/tauri-commands/dock.ts
// macOS Dock (whether Cmdr may be offered a tile, and putting it there)

import { commands } from '$lib/ipc/bindings'
import type { DockPinFailure, DockPinState } from '$lib/ipc/bindings'
import { getAppLogger } from '$lib/logging/logger'

const log = getAppLogger('dock')

/**
 * Whether Cmdr may be offered a place in the Dock, and whether it's already there.
 *
 * One value covers "no tile yet", "already pinned", and every reason to stay quiet, so no caller
 * has to re-derive where Cmdr is installed. See `src-tauri/src/dock/CLAUDE.md`.
 *
 * A backend that can't answer at all reads as `preferencesUnreadable`, which keeps the nudge
 * silent rather than firing it on a guess.
 */
export async function getDockPinState(): Promise<DockPinState> {
  try {
    return await commands.getDockPinState()
  } catch (e) {
    log.warn(`Couldn't ask about the Dock: ${String(e)}`)
    return { kind: 'unavailable', reason: 'preferencesUnreadable' }
  }
}

/**
 * Adds Cmdr to the Dock as the leftmost app tile and restarts the Dock so it appears.
 *
 * Answers `null` on success, or the typed reason it didn't happen. The Dock visibly blinks for
 * about a second while it reloads; that's the restart, and it's expected.
 *
 * An IPC that never comes back at all lands on `timedOut`, which is what it means to the caller:
 * nobody knows whether the tile made it.
 */
export async function addCmdrToDock(): Promise<DockPinFailure | null> {
  try {
    const res = await commands.addCmdrToDock()
    return res.status === 'error' ? res.error : null
  } catch (e) {
    log.warn(`The Dock pin never answered: ${String(e)}`)
    return { kind: 'timedOut' }
  }
}

File: apps/desktop/src-tauri/src/commands/dock.rs
//! IPC commands over the macOS Dock.
//!
//! Pass-throughs: every decision about whether Cmdr may be offered a Dock tile, and what tile to
//! write, lives in `../dock/`. See `../dock/CLAUDE.md`.

use tokio::time::Duration;

use crate::commands::util::{blocking_typed_result_with_timeout, blocking_with_timeout};
use crate::dock::{DockPinBlocker, DockPinFailure, DockPinState};

/// Reading the Dock's preferences goes through `cfprefsd` over XPC, which can wait on a busy
/// daemon, so it carries the standard 2 s read deadline.
const READ_TIMEOUT: Duration = Duration::from_secs(2);

/// The pin writes preferences and then waits for the Dock to acknowledge a restart, so it gets the
/// 5 s write tier rather than the read one.
const WRITE_TIMEOUT: Duration = Duration::from_secs(5);

/// Whether we may offer to put Cmdr in the Dock, and whether it's already there.
///
/// A slow or unreachable `cfprefsd` answers `PreferencesUnreadable`, which keeps the nudge silent
/// rather than firing on a guess.
#[tauri::command]
#[specta::specta]
pub async fn get_dock_pin_state() -> DockPinState {
    blocking_with_timeout(
        READ_TIMEOUT,
        DockPinState::Unavailable {
            reason: DockPinBlocker::PreferencesUnreadable,
        },
        crate::dock::pin_state,
    )
    .await
}

/// Adds Cmdr to the Dock as the first tile and restarts the Dock so it appears.
///
/// Only ever called after the person says yes to the nudge. It re-checks everything
/// [`get_dock_pin_state`] checked, so a stale answer can't turn into a write we'd never have
/// offered.
#[tauri::command]
#[specta::specta]
pub async fn add_cmdr_to_dock() -> Result<(), DockPinFailure> {
    blocking_typed_result_with_timeout(
        WRITE_TIMEOUT,
        || DockPinFailure::TimedOut,
        |e| {
            log::warn!(target: "dock", "The Dock pin task didn't finish: {e}");
            DockPinFailure::TimedOut
        },
        crate::dock::pin_cmdr,
    )
    .await
}
